ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -0,0 +1,221 @@
+    ) {
+        ModelTensors tensors;
+        if (Strings.isBlank(body)) {
+            log.error("Remote model response body is empty!");
+            tensors = processErrorResponse(statusCode, "null");
+        } else {
+            if (statusCode < HttpStatus.SC_OK || statusCode > HttpStatus.SC_MULTIPLE_CHOICES) {
+                log.error("Remote server returned error code: {}", statusCode);
+                tensors = processErrorResponse(statusCode, body);
+            } else {
+                try {
+                    tensors = processOutput(body, connector, scriptService, parameters);
+                } catch (Exception e) {
+                    log.error("Failed to process response body: {}", body, e);
+                    tensors = processErrorResponse(statusCode, body);
+                }
+            }
+        }
+        tensors.setStatusCode(statusCode);
+        tensorOutputs.put(countDownLatch.getSequence(), tensors);
+    }
+
+    private void reOrderTensorResponses(Map<Integer, ModelTensors> tensorOutputs) {
+        List<ModelTensors> modelTensors = new ArrayList<>();
+        TreeMap<Integer, ModelTensors> sortedMap = new TreeMap<>(tensorOutputs);
+        log.debug("Reordered tensor outputs size is {}", sortedMap.size());
+        if (tensorOutputs.size() == 1) {
+            // batch API case
+            int status = tensorOutputs.get(0).getStatusCode();
+            if (status == HttpStatus.SC_OK) {
+                modelTensors.add(tensorOutputs.get(0));
+                actionListener.onResponse(modelTensors);
+            } else {
+                try {
+                    actionListener
+                        .onFailure(
+                            new OpenSearchStatusException(
+                                AccessController
+                                    .doPrivileged(
+                                        (PrivilegedExceptionAction<String>) () -> GSON
+                                            .toJson(tensorOutputs.get(0).getMlModelTensors().get(0).getDataAsMap())
+                                    ),
+                                RestStatus.fromCode(status)
+                            )
+                        );
+                } catch (PrivilegedActionException e) {
+                    actionListener.onFailure(new OpenSearchStatusException(e.getMessage(), RestStatus.fromCode(statusCode)));
+                }
+            }
+        } else {
+            // non batch API.
+            for (Map.Entry<Integer, ModelTensors> entry : sortedMap.entrySet()) {
+                modelTensors.add(entry.getKey(), entry.getValue());
+            }
+            actionListener.onResponse(modelTensors);
+        }
+    }
+
+    protected class MLResponseSubscriber implements Subscriber<ByteBuffer> {
+        private Subscription subscription;
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            this.subscription = s;
+            s.request(Long.MAX_VALUE);
```

Comment:
What does this mean? what happens if we set min value?

Reply:
This is a definition which tells the publisher how many at most the subscriber can fetch in one time, you can find more here: https://www.reactive-streams.org/reactive-streams-1.0.0-javadoc/org/reactivestreams/Subscription.html. The value must be a positive number and if we set it to a small number it still works but internally the publisher will publish more times until all data are sent to subscriber. Set it to a bigger value can make this process more efficient.